.github/workflows/test.yml

```
@@ -5,6 +5,12 @@ jobs:
   test:
     name: Run tests
     runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        node-version:
+          - 12.17.0
+          - 14.13.1
+          - 16.0.0
```

It's unnecessary to target versions. It will only test the specific versions that you provide. Then you will miss 12.18.0, 14.14.0, 16.1.0, 17.0.0 and others.

We can target versions like this:

```suggestion
        node-version:
          - ^12.17
          - ^14.13.1
          - ^16
          - ^17
```

---

BTW, there is no difference between `^12.17, ^14.13.1, ^16, ^17` and `12, 14, 16, 17`. `setup-node` will pick the latest version from the scope that you provide. Both are OK.

---

Not sure if I'm 100% in agreement with you here. I want to test on specifically the lowest versions `package.json` says we're supporting because if I don't I can't be sure those actually work. E.g. I may start using a language feature introduced in `14.14.0`, which would break for users complying with the `engines` field, but I wouldn't know it until after the changes was released.

You could try to argue the same for higher versions, but we have to assume Node is doing it's job correctly and any version `^14.13.1` works if `14.13.1` works.

That being said, I completely agree we should include a v17 here (latest seems fine to me). As for v16, on the one hand that "feels" fine to me to test both `16.0.0` and `^16`. On the other hand, to be consistent we should then also test `12.17.0` and `^12` as well as `14.13.1` and `^14`.

Given all that, I'd instead suggest either:

```suggestion
        node-version:
          # Lowest versions we say we support (see package.json)
          - 12.20.0
          - 14.13.1
          - 16.0.0
          # Highest version we're interested in
          - ^17
```

Or:

```suggestion
        node-version:
          # Lowest versions we say we support (see package.json)
          - 12.20.0
          - 14.13.1
          - 16.0.0
          # Highest versions we're interested in
          - ^12
          - ^14
          - ^16
          - ^17
```

(with 8db814737b06d8d9095fdef5bf1850b50eada4a6, the v12 test target should be updated to `12.20.0`)

---

`12.20.0, 14.13.1, 16.0.0, 12, 14, 16, 17` looks good. `^` is unnecessary but it's up to you.